flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/schema/FlinkTable.scala

```diff
@@ -44,14 +44,14 @@ abstract class FlinkTable[T](
 
   val fieldTypes: Array[TypeInformation[_]] =
     typeInfo match {
-      case cType: CompositeType[T] =>
+      case cType: CompositeType[_] =>
```

Comment:
The fieldTypes can be get from `TableEnvironment.getFieldTypes(typeInfo)` and the length validation could be placed above.